libshpool/src/config_watcher.rs

```diff
@@ -0,0 +1,676 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Context;
+use crossbeam_channel::{select, unbounded, Receiver, RecvError, Sender};
+use notify::{
+    event::ModifyKind, recommended_watcher, Event, EventKind, INotifyWatcher, RecursiveMode,
+    Watcher as _,
+};
+use std::{
+    collections::{hash_map::Entry, HashMap},
+    mem,
+    path::{Path, PathBuf},
+    thread::{self, JoinHandle},
+    time::{Duration, Instant},
+};
+use tracing::{debug, error};
+
+/// Watches on `path`, returnes the watched path, which is the closest existing ancestor of `path`, and
+/// the immediate child that is of interest.
+pub fn best_effort_watch<'a>(
+    watcher: &mut INotifyWatcher,
+    path: &'a Path,
+) -> anyhow::Result<(&'a Path, Option<&'a Path>)> {
+    let mut watched_path = Err(anyhow::anyhow!("empty path"));
```

Comment:
elsewhere in the program I always import the anyhow macro instead of fully qualifying, let's stick to that pattern.